File: core/src/jvmMain/java/de/haukesomm/sokoban/core/moving/validation/PreconditionMoveValidator.java
package de.haukesomm.sokoban.core.moving.validation;

import de.haukesomm.sokoban.core.Direction;
import de.haukesomm.sokoban.core.Entity;
import de.haukesomm.sokoban.core.GameState;
import de.haukesomm.sokoban.core.moving.MoveValidatorStatus;

import java.util.Collection;

public class PreconditionMoveValidator extends AbstractMoveValidator {
    private final MoveValidator preconditionMoveValidator;
    private final Collection<MoveValidator> moveValidators;

    public PreconditionMoveValidator(
            MoveValidator preconditionMoveValidator,
            Collection<MoveValidator> moveValidators
    ) {
        this.preconditionMoveValidator = preconditionMoveValidator;
        this.moveValidators = moveValidators;
    }

    @Override
    public Collection<MoveValidatorStatus> check(GameState state, Entity entity, Direction direction) {
        var statuses = preconditionMoveValidator.check(state, entity, direction);
        if (statuses.contains(MoveValidatorStatus.IMPOSSIBLE)) {
            return statuses;
        }

        moveValidators.stream()
                .map(validator -> validator.check(state, entity, direction))
                .forEach(statuses::addAll);

        return statuses;
    }
}

File: core/src/commonMain/kotlin/de/haukesomm/sokoban/core/moving/validation/PreconditionMoveValidator.kt
package de.haukesomm.sokoban.core.moving.validation

import de.haukesomm.sokoban.core.Direction
import de.haukesomm.sokoban.core.Entity
import de.haukesomm.sokoban.core.state.GameState

class PreconditionMoveValidator(
    private val preconditionMoveValidator: MoveValidator,
    private val moveValidators: Collection<MoveValidator>
) : AbstractMoveValidator() {

    override fun check(state: GameState, entity: Entity, direction: Direction): Collection<MoveValidatorStatus> {
        val statuses = preconditionMoveValidator.check(state, entity, direction).toMutableSet()
        if (statuses.contains(MoveValidatorStatus.IMPOSSIBLE)) {
            return statuses
        }

        moveValidators
            .map { it.check(state, entity, direction) }
            .forEach { statuses += it }

        return statuses
    }
}
